_posts/2018-09-06-stateless-microservices.md

```diff
@@ -0,0 +1,304 @@
+
+## Why are we doing this now?
+
+There is a lot of overlap between modern, cloud-native microservices and FaaS functions as I'll go on to explain in the next section. OpenFaaS has always had support for running any container or process whether that be a FaaS Function, the AWS CLI, ImageMagick, or even PowerShell on Windows. Two requests came in from the community within a short period of time which acted as a catalyst for this next step in our story.
+
+One of our newest users at [Wireline.io](https://www.wireline.io) raised a [feature request](https://github.com/openfaas/faas/issues/716) to enhance the HTTP route functionality of functions. Wireline wanted to write functions which could run both on [AWS Lambda](https://aws.amazon.com/free/webapps/?https://aws.amazon.com/lambda/) and on OpenFaaS without any additional changes. Around that same time the CEO of [GitLab](https://about.gitlab.com/), [Sid Sijbrandi](https://twitter.com/sytses?lang=en) reached out to learn more about Serverless and how it could be leveraged at GitLab. Sid asked me whether OpenFaaS could be used to manage both FaaS Functions and the microservices his team was more familiar with such as a Sinatra apps. He showed a particular interest in [scaling to zero](https://www.openfaas.com/blog/zero-scale/) when idle.
+
+Let's look at some background before moving into the tutorial showing how to put this into practice.
+
+## What's a Function?
+
+Before getting into what a "stateless microservice" is, let's go back to my definition of a "FaaS Function" as defined in my blog post: [Functions as a Service (FaaS) (Jan 2017)](https://blog.alexellis.io/functions-as-a-service/).
+
+Functions tend to involve:
+
+* invoking short-lived functions (Lambda has a default 1-sec timeout)
+* does not publish TCP services - often accesses third-party services or resources
+* tends to be ad-hoc/event-driven such as responding to webhooks
+* should deal gracefully with spikes in traffic
+* despite name, runs on servers backed by a fluid, non-deterministic infrastructure
+* when working well makes the complex appear simple
+* has several related concerns: infrastructure management, batching, access control, definitions, scaling & dependencies
+
+Since the original post I re-wrote my initial observations as a series of properties.
+
+Functions are:
+
+* stateless
+* emphemeral & short-lived
+* auto-scaling
+* single-purposed
+
+Functions are stateless because they do not rely on internal state-machines, stored files or sessions. Each call to the function should result in the same end-result. They don't have to be strictly idempotent, but they are easier to manage when they are.
```

Comment:
Slight difficulty with the sentence _Each call to the function should result in the same end result_, when applied to the CRUD example below which can clearly return a different value eg after update.

I think clarify somehow that the function as deployed has no state, but it can obviously reference external state eg in a db and therefore deliver varying results?